rust/chains/tw_polkadot/src/scale.rs

```diff
@@ -0,0 +1,442 @@
+/// TODO: this is a substrate-specific encoding, but consider moving to tw_encoding crate
+///
+
+pub trait ToScale {
+    fn to_scale(&self) -> Vec<u8> {
+        let mut data = Vec::new();
+        self.to_scale_into(&mut data);
+        data
+    }
+
+    fn to_scale_into(&self, out: &mut Vec<u8>);
+}
+
+impl ToScale for bool {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        out.push(if *self { 0x01 } else { 0x00 } as u8);
+    }
+}
+
+macro_rules! fixed_impl {
+    ($($t:ty),+) => {
+        $(impl ToScale for $t {
+            fn to_scale_into(&self, out: &mut Vec<u8>) {
+                out.extend_from_slice(&self.to_le_bytes())
+            }
+        })+
+    };
+}
+
+fixed_impl!(u8, u16, u32, u64, usize, i8, i16, i32, i64, isize);
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub struct Compact<T>(pub T);
+
+// Implementations for Compact
+
+impl ToScale for Compact<u8> {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        match self.0 {
+            0..=0b0011_1111 => out.push(self.0 << 2),
+            _ => (((self.0 as u16) << 2) | 0b01).to_scale_into(out),
+        }
+    }
+}
+
+impl ToScale for Compact<u16> {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        match self.0 {
+            0..=0b0011_1111 => out.push((self.0 as u8) << 2),
+            0..=0b0011_1111_1111_1111 => ((self.0 << 2) | 0b01).to_scale_into(out),
+            _ => (((self.0 as u32) << 2) | 0b10).to_scale_into(out),
+        }
+    }
+}
+
+impl ToScale for Compact<u32> {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        match self.0 {
+            0..=0b0011_1111 => out.push((self.0 as u8) << 2),
+            0..=0b0011_1111_1111_1111 => (((self.0 as u16) << 2) | 0b01).to_scale_into(out),
+            0..=0b0011_1111_1111_1111_1111_1111_1111_1111 => {
+                ((self.0 << 2) | 0b10).to_scale_into(out)
+            },
+            _ => {
+                out.push(0b11);
+                self.0.to_scale_into(out);
+            },
+        }
+    }
+}
+
+impl ToScale for Compact<u64> {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        match self.0 {
+            0..=0b0011_1111 => out.push((self.0 as u8) << 2),
+            0..=0b0011_1111_1111_1111 => (((self.0 as u16) << 2) | 0b01).to_scale_into(out),
+            0..=0b0011_1111_1111_1111_1111_1111_1111_1111 => {
+                (((self.0 as u32) << 2) | 0b10).to_scale_into(out)
+            },
+            _ => {
+                let bytes_needed = 8 - self.0.leading_zeros() / 8;
+                out.reserve(bytes_needed as usize);
+                out.push(0b11 + ((bytes_needed - 4) << 2) as u8);
+                let mut x = self.0;
+                for _ in 0..bytes_needed {
+                    out.push(x as u8);
+                    x >>= 8;
+                }
+            },
+        }
+    }
+}
+
+// TODO: implement U128 support (we don't have it yet in tw_number)
```

Comment:
u128 and i128 are standard Rust types, but it would be useful to implement some functions that we have for U256, I256